### PR TITLE
add support for ShellyTRV

### PR DESCRIPTION
**Pull request by hinrikus**
WIP
needs still a prettier interface

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -131,6 +131,22 @@
...
+         <template id="template-shellytrv">
+            <div class="thing thing-trv">
+                <div class="thing-detail">
+                    <div class="edit"> <i class="feather" data-feather="edit"></i> </div>
+                    <!--<input type="checkbox">-->
+                    <div class="thing-value">
+                        <span name="value">-</span> °C
+                        <div>
+                            <span name="plus" ><i class="feather-big" data-feather="plus"></i></span>
+                            <span name="minus" ><i class="feather-big" data-feather="minus"></i></span>
```
Comment (trilader) [inline index.html]: There is a space before the closing bracket on the outer span tag that can be removed.

File: js/main.js [hunk 2]
```diff
@@ -132,7 +133,7 @@ let checkbox_initializer = (thing, e) => {
             id: thing.id,
             value: checkbox.checked,
         }));
-        //div.classList.remove(checkbox.checked ? "off": "on");
+        //div.classList.remove(checkbox.checkboxchecked ? "off": "on");
```
Comment (trilader) [inline js/main.js]: Why is this change here? Looks like a not quite correct search/replace?

File: js/main.js [hunk 3]
```diff
@@ -146,9 +147,47 @@ let checkbox_initializer = (thing, e) => {
...
+    let div = e.querySelector("div.thing-detail");
+    let plus = e.querySelector("span[name='plus']");
+    let minus = e.querySelector("span[name='minus']");
+    let value_ele = e.querySelector("span[name='value']");
+    let new_value;
+    plus.addEventListener('click', () => {
+        //plus.click()
+        div.classList.add("pending");
+        new_value = parseInt(value_ele.innerText) + 1
+        console.log(new_value);
```
Comment (trilader) [inline js/main.js]: Do we need the `console.log`?

File: js/main.js [hunk 4]
```diff
@@ -146,9 +147,47 @@ let checkbox_initializer = (thing, e) => {
...
+        cb()
+    });
+    minus.addEventListener('click', () => {
+        //minus.click();
+        div.classList.add("pending");
+        new_value = parseInt(value_ele.innerText) - 1;
+        console.log(value_ele.innerText);
```
Comment (trilader) [inline js/main.js]: Do we need the `console.log`?

File: js/main.js [hunk 5]
```diff
@@ -146,9 +147,47 @@ let checkbox_initializer = (thing, e) => {
...
+            type: "command",
+            id: thing.id,
+            value: new_value,
+        }));
+        add_pending_change(thing, setTimeout(() => {
+            value_ele.innerText = new_value;
+            div.classList.remove("pending");
+        }, 1000));
+        div.classList.remove("pending");
+    };
```
Comment (trilader) [inline js/main.js]: It would be better to move the definition above the usages in the plus and minus button event handlers above. It confused me and I first thought this code was calling the `cb` defined above for a different kind of handler.

File: js/main.js [hunk 6]
```diff
@@ -146,9 +147,47 @@ let checkbox_initializer = (thing, e) => {
...
+        //minus.click();
+        div.classList.add("pending");
+        new_value = parseInt(value_ele.innerText) - 1;
+        console.log(value_ele.innerText);
+        cb()
+    });
+
+    let cb = () => {
+
+        console.log(new_value);
```
Comment (trilader) [inline js/main.js]: Do we need that `console.log` and empty line above?

File: js/main.js [hunk 7]
```diff
@@ -146,9 +147,47 @@ let checkbox_initializer = (thing, e) => {
...
+        socket.send(JSON.stringify({
+            type: "command",
+            id: thing.id,
+            value: new_value,
+        }));
+        add_pending_change(thing, setTimeout(() => {
+            value_ele.innerText = new_value;
```
Comment (trilader) [inline js/main.js:201]: This would set the element to the new value regardless of success or not.

File: js/main.js [hunk 8]
```diff
@@ -643,7 +682,7 @@ let ws_connect = () => {
         const l = window.location;
         const c = config;
         return (c.ws_proto || l.protocol.replace("http", "ws")) + "//"
-            + (c.ws_host || l.host)
+            + (c.ws_host || l.hostname)
```
Comment (trilader) [inline js/main.js:722]: Using `Location.hostname` is indeed correct as it doesn't include the port. Please move this change to a separate commit.

**Review by trilader — CHANGES_REQUESTED** [review]

File: js/main.js (hunk 3 above)
Comment (hinrikus) [reply js/main.js]: Nope.

File: js/main.js (hunk 2 above)
Comment (hinrikus) [reply js/main.js]: I have no idea. I would remove these lines since they are commented out

File: js/main.js (hunk 7 above)
Comment (hinrikus) [reply js/main.js:201]: yes. this behaviour is analogue to the behaviour of a button press. The button will change its colour if we press it and keep that colour when the confirmation comes from websocket.
Would you suggest other behaviour?

Comment (trilader) [reply js/main.js:201]: The issue isn't with the pattern but the content of the function within the `setTimeout`. It will be called when the timeout for receiving an answer from the backend expires, which will currently set the text of the label to `new_value` (which is what we wanted to set it to if it had succeeded) instead of the previous value (which is still the current value as the server didn't respond in time when the timeout callback gets called).

File: js/main.js (hunk 2 above)
Comment (trilader) [reply js/main.js]: Ok

File: js/main.js (hunk 7 above)
Comment (hinrikus) [reply js/main.js:201]: ok. how would you fix this?

Comment (trilader) [reply js/main.js:201]: We can store `value_ele.innerText` in a variable before we set a the new value. The callback can use the stored value to restore `value_ele.innerText` if the `setTimeout` expires.
